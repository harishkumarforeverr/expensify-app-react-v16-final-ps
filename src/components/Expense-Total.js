export default (expenses)=>expenses
                                .map((expense)=>parseFloat(expense.amount,10))
                                .reduce((sum,value)=>(sum+value),0);