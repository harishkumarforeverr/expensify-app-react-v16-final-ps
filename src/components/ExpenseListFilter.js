import React from "react";
import {connect} from "react-redux";
import {DateRangePicker} from "react-dates";
import moment from "moment";
import {setTextFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from "../actions/Filter";

export class ExpenseListFilter extends React.Component{
    constructor(props){  
        super(props);
      this.state={ 
        focusedInput:null
      }
    }
    onDatesChange=({startDate,endDate})=>{ 
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onSearchInputChange=(e)=>{ 
        this.props.setTextFilter(e.target.value);
    }
    onSelectedTagChange=(e)=>{
        const text=e.target.value;
        if(text==='amount'){
         this.props.sortByAmount();
        }else if(text=="date"){
            this.props.sortByDate();
        }
    }
   render(){
       return (
        <div> 
            <input
            type="text"
            placeholder="search" 
            value ={this.props.filters.text} 
            name="textInput"
            onChange={this.onSearchInputChange}
            />
            <select 
            value={this.props.filters.sortBy}
             onChange={this.onSelectedTagChange}  >
                <option value="date">Date</option> 
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker 
             startDate={this.props.filters.startDate}
             startDateId="startDateId"
             endDate={this.props.filters.endDate}
             endDateId="endDateId"
             onDatesChange={this.onDatesChange}
             focusedInput={this.state.focusedInput}
             onFocusChange={focusedInput=>this.setState({ focusedInput })}
             numberOfMonths={1}
             isOutsideRange={()=>false}
             displayFormat="DD/MM/YYYY"
             showClearDates={true}
            />

        </div>
    )
   }
}
 
const mapStateToProps=(state)=>({
    filters:state.filters 
})
 
const mapPropsToState=(dispatch)=>({
    setStartDate:(startDate)=>dispatch(setStartDate(startDate)),
    setEndDate:(endDate)=>dispatch(setEndDate(endDate)), 
    sortByDate:()=>dispatch(sortByDate()),
    sortByAmount:()=>dispatch(sortByAmount()),
    setTextFilter:(text)=>dispatch(setTextFilter(text)) 
})
export default connect(mapStateToProps,mapPropsToState)(ExpenseListFilter);