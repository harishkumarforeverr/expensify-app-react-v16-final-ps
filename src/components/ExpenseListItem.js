import React from 'react'; 
import moment from "moment";
import {Link} from "react-router-dom";
import formatAmount from 'indian-currency-formatter';

const ExpenseListItem=({description,amount,createdAt,id })=>{
    return (
        <div> 
           <Link to={`edit/${id}`}>
           <h3>{description} </h3> 
           <p>  
              &#x20B9;{formatAmount(parseFloat(amount,10)) } 
              - 
              {moment(createdAt).format("Do MMM, YYYY")}  
           </p> 
           </Link> 
        </div>
    )
} 

export default ExpenseListItem;