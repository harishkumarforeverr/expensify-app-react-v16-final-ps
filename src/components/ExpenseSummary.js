import React from "react";
import {connect} from "react-redux";
import formatAmount from 'indian-currency-formatter';
import ExpensesTotal from "./Expense-Total"; 
import visibleExpense from "../VisibleExpenseSelector/Expense";

export const ExpenseSummary=(props)=>{
    const expensesWord= props.noOfExpenses==1?"expense":"expenses";
    return (
        <div>
          {
              (props.noOfExpenses!==0) &&
              <h1>Viewing {props.noOfExpenses} {expensesWord} of totaling &#x20B9;{formatAmount(props.TotalExpenses)}  </h1>
          }            
        </div>
    )
}
 
const mapStateToProps=(state)=>{
    const expenses=visibleExpense(state.expenses,state.filters); 
    return {
        noOfExpenses:expenses.length,
        TotalExpenses:ExpensesTotal(expenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);
