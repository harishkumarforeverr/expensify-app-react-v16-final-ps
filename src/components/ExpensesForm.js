import React from 'react';
import "react-dates/initialize";
import {SingleDatePicker} from "react-dates";
import moment from "moment";
 
export default class ExpensesForm extends React.Component{
    constructor(props){
        super(props)
        this.state={
            description:props.description?props.description:"",
            note:props.note?props.note:"",
            amount:props.amount?parseFloat(props.amount,10):"",
            createdAt:props.createdAt?moment(props.createdAt):moment(),
            error:undefined,
            focused:false
        }
    }
  onDescriptionChange=(e)=>{
    const description=e.target.value;
     this.setState(()=>({ 
         description
       }))
  }      
  onNoteChange=(e)=>{
    const note=e.target.value;
     this.setState(()=>({ 
         note
       }))
  }
  onAmountChange=(e)=>{
    const amount=e.target.value;
    if(amount.match(/^\d{1,}(\.\d{0,2})?$/) || !amount){
        this.setState(()=>({ 
            amount
          }))
    }
  }
  onSubmit=(e)=>{
        e.preventDefault(); 
        let error;
        if(!this.state.description.trim() || !this.state.amount){ 
            error="please enter amount and description";
            this.setState(()=>({ error }));
        }else{
          this.props.onSubmit({
              description:this.state.description,
              amount:this.state.amount,
              note:this.state.note,
              createdAt:this.state.createdAt
          })   
          this.setState(()=>({ error }));
        }
  }
  onDateChange=(createdAt)=>{
    if(createdAt){
        this.setState({createdAt})
    }
  }
  render(){
      return (
          <div> 
              {this.state.error && <p>{this.state.error}</p>}
              <form onSubmit={this.onSubmit}>
              <input 
                value={this.state.description}
                placeholder="Add a new Expense"
                type="text"
                name="add_input"
                autoFocus
                onChange={this.onDescriptionChange}
                /> 
                <input 
                value={this.state.amount}
                placeholder="amount"
                type="number"
                name="amount_input"
                onChange={this.onAmountChange}
                /> 
                <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.focused}
                onFocusChange={({focused})=> this.setState({focused}) }
                numberOfMonths={1}
                isOutsideRange={()=>false}
                displayFormat="DD/MM/YYYY"
                id="singleDatePicker_id"
                />
                <textarea
                value={this.state.note}
                placeholder="Add notes to your expenses(Optional !)"
                name="amount_input"
                onChange={this.onNoteChange}
                > 
                </textarea>
                <button> Add Expense </button>                 
              </form>
          </div>
      )
  }
}