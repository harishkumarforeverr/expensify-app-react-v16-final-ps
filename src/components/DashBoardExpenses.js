import React from "react"
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
const DashBoardExpenses=()=>{
    return (
        <div> 
         <ExpenseListFilter />
         <ExpenseList />
        </div>
    )
}
export default DashBoardExpenses;