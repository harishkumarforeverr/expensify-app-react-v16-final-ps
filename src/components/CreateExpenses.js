import React from 'react'
import ExpensesForm from "./ExpensesForm";
import {connect} from "react-redux";
import {addExpense} from "../actions/Expense"
export const CreateExpenses=(props)=>{ 
    return (
        <div>
         <h1> Add Expense </h1>
         <ExpensesForm onSubmit={(expense)=>{
            props.addExpense(expense)
            props.history.push("/");
         }}/>
        </div>
    )
}

const mapDispatchToState=(dispatch)=>{
 return {
     addExpense:(expense)=>dispatch(addExpense(expense))
 }
}

export default connect(undefined,mapDispatchToState)(CreateExpenses);