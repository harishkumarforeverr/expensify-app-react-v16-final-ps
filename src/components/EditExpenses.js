import React from 'react';
import {connect} from "react-redux";
import ExpensesForm from "./ExpensesForm";
import {editExpense,removeExpense} from "../actions/Expense";

export const EditExpenses=(props)=>{    
    return (
        <div> 
            <h1>Edit Expense </h1>
             <ExpensesForm 
             {...props.editingExpense}
             onSubmit={(expense)=>{ 
                 props.editExpense(props.editingExpense.id,expense)
                 props.history.push("/");
             }}
              />   
           <button onClick={()=>{
               props.removeExpense({id:props.editingExpense.id})
               props.history.push("/");
               }} >Remove</button>
        </div>
    )
}

const mapStateToProps=(state,props)=>{
    return {
        editingExpense:state.expenses.find((expense)=>expense.id===props.match.params.id)
    }
}

const mapDispatchToProps=(dispatch)=>{
    return {
        editExpense:(id,expense)=>dispatch(editExpense(id,expense)),
        removeExpense:({id})=>dispatch(removeExpense({id}))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpenses);