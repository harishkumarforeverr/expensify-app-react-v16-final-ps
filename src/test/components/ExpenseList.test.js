import {shallow} from 'enzyme';
import React from 'react';
import expenses from '../fixtures/expenses';
import {ExpenseList} from "../../components/ExpenseList";

test('import {ExpenseList} from "../../components/ExpenseList"',()=>{
    const wrapper=shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot();
})

test('import {ExpenseList} from "../../components/ExpenseList with no expenses"',()=>{
     const wrapper=shallow(<ExpenseList expenses={[]} />)
     expect(wrapper).toMatchSnapshot();
})