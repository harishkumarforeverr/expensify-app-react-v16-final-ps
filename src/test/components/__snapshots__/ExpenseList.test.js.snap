// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import {ExpenseList} from "../../components/ExpenseList with no expenses" 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`import {ExpenseList} from "../../components/ExpenseList" 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gun"
    id={0}
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={"1969-12-28T00:00:00.000Z"}
    description="rent"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={"1970-01-05T00:00:00.000Z"}
    description="credit Card"
    id={2}
    key="2"
    note=""
  />
</div>
`;
